### PR TITLE
Add initial scaffold for portal lc network

### PR DESCRIPTION
**Pull request by KonradStaniec**
Add:
- Initial scaffold for light client network according to current spec (which is influx)
- Add initial handling of bootstrap object

### CODE REVIEW

File: fluffy/network/light_client/light_client_content.nim [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+type
+  ContentType* = enum
+    lightClientBootstrap = 0x00
+    lightClientUpdate = 0x01
+    lightClientFinalityUpdate = 0x02
+    lightClientOptimisticUpdate = 0x03
+
+  # TODO Consider how we will gossip boostraps?
+  # In normal LC operation node trust only one offered bootstrap, therefore offers
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:28]: Has this remark been made yet on the specs repo? If not, it's a good point so might want to add it there.

File: fluffy/network/light_client/light_client_content.nim [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+export ssz_serialization, common_types, hash
+
+type
+  ContentType* = enum
+    lightClientBootstrap = 0x00
+    lightClientUpdate = 0x01
+    lightClientFinalityUpdate = 0x02
+    lightClientOptimisticUpdate = 0x03
+
+  # TODO Consider how we will gossip boostraps?
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:27]:
```suggestion
  # TODO Consider how we will gossip bootstraps?
```

File: fluffy/network/light_client/light_client_content.nim [hunk 3]
```diff
@@ -0,0 +1,109 @@
+# Nimbus
+# Copyright (c) 2021-2022 Status Research & Development GmbH
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:2]:
```suggestion
# Copyright (c) 2022 Status Research & Development GmbH
```

File: fluffy/network/light_client/light_client_content.nim [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      lightClientUpdateKey*: LightClientUpdateKey
+    of lightClientFinalityUpdate:
+      lightClientFinalityUpdateKey*: LightClientFinalityUpdateKey
+    of lightClientOptimisticUpdate:
+      lightClientOptimisticUpdateKey*: LightClientOptimisticUpdateKey
+
+  # Interal object to prepend forkdigest to bootstrap
+  ForkedLCBootsrap = object
+    fd: ForkDigest
+    bs: altair.LightClientBootstrap
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:54]: LightClient is written out fully for all other objects, lets do it here too. Also, type. Same for others.
```suggestion
  ForkedLightClientBootstrap = object
    forkDigest: ForkDigest
    bootstrap: altair.LightClientBootstrap
```

File: fluffy/network/light_client/light_client_content.nim [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+proc decodeBootstrap(
+    data: openArray[byte]): Result[altair.LightClientBootstrap, string] =
+  try:
+    let decoded = SSZ.decode(
+      data,
+      altair.LightClientBootstrap
+    )
+    return ok(decoded)
+  except MalformedSszError as exc:
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:81]: For any SSZ decoding I think it is fine to just use the covering `SszError`

File: fluffy/network/light_client/light_client_content.nim [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    of lightClientBootstrap:
+      lightClientBootstrapKey*: LightClientBootstrapKey
+    of lightClientUpdate:
+      lightClientUpdateKey*: LightClientUpdateKey
+    of lightClientFinalityUpdate:
+      lightClientFinalityUpdateKey*: LightClientFinalityUpdateKey
+    of lightClientOptimisticUpdate:
+      lightClientOptimisticUpdateKey*: LightClientOptimisticUpdateKey
+
+  # Interal object to prepend forkdigest to bootstrap
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:51]:
```suggestion
  # Internal object to prepend forkdigest to bootstrap
```

Except that it is not really an internal only object? It is the object that is required to send over the wire?

File: fluffy/network/light_client/light_client_content.nim [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    return err(exc.msg)
+  except SszSizeMismatchError as exc:
+    return err(exc.msg)
+
+proc encodeBootstrapForked*(
+    fork: ForkDigest,
+    bs: altair.LightClientBootstrap): seq[byte] =
+  SSZ.encode(ForkedLCBootsrap(fd: fork, bs: bs))
+
+proc decodeBootstrapForked*(
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_content.nim:91]: Fine for now, but I noticed that the specs are written wrong / differently for this: `SSZ.serialize(ForkDigest + LightClientBootstrap)`. Not sure if the intention in specs is / was, prepend before or after SSZ serialization.

File: fluffy/network/light_client/light_client_network.nim [hunk 8]
```diff
@@ -0,0 +1,167 @@
+# Nimbus
+# Copyright (c) 2021-2022 Status Research & Development GmbH
```
Comment (kdeme) [inline fluffy/network/light_client/light_client_network.nim:2]:
```suggestion
# Copyright (c) 2022 Status Research & Development GmbH
```

File: fluffy/tests/test_light_client_content.nim [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# Nimbus
+# Copyright (c) 2021-2022 Status Research & Development GmbH
```
Comment (kdeme) [inline fluffy/tests/test_light_client_content.nim:2]:
```suggestion
# Copyright (c) 2022 Status Research & Development GmbH
```

**Review by kdeme — CHANGES_REQUESTED** [review]

File: fluffy/network/light_client/light_client_content.nim (hunk 6 above)
Comment (KonradStaniec) [reply fluffy/network/light_client/light_client_content.nim:51]: Hmm maybe I phrased it ambiguously. It is internal to in sense as internal to this nim module. Other nim modules does not need to know about this object.

File: fluffy/network/light_client/light_client_content.nim (hunk 4 above)
Comment (KonradStaniec) [reply fluffy/network/light_client/light_client_content.nim:54]: sure, it just sounded mouthful

File: fluffy/network/light_client/light_client_content.nim (hunk 7 above)
Comment (KonradStaniec) [reply fluffy/network/light_client/light_client_content.nim:91]: I will add a question there then, to be sure.

Effect is ultimately the same for decoder, as `array[4, byte]`  is ssz serialized to static 4 bytes.

**Pull request merged**